sklearn/manifold/tests/test_t_sne.py

```diff
@@ -890,3 +890,81 @@ def test_tsne_n_jobs(method):
                 angle=0, n_jobs=2, random_state=0).fit_transform(X)
 
     assert_allclose(X_tr_ref, X_tr)
+
+
+def test_tsne_with_mahalanobis_distance():
+    """Make sure that tha mahalanobis distance works with metric_params
+       properly set and it doesn't otherwise"""
+    random_state = check_random_state(0)
+    n_features = 10
+    n_embedding = 3
+    n_samples = 500
+    X = random_state.randn(n_samples, n_features)
+
+    # 1. raises error here (original issue)
+    tsne = TSNE(verbose=1, perplexity=40, n_iter=250, learning_rate=50,
+                n_components=n_embedding, random_state=0, metric='mahalanobis')
+    ref = "Must provide either V or VI for Mahalanobis distance"
+    with pytest.raises(ValueError, match=ref):
+        tsne.fit_transform(X)
+
+    # 2. check for correct answer
+    precomputed_X = squareform(pdist(X, metric='mahalanobis'), checks=True)
+    ref = TSNE(verbose=1, perplexity=40, n_iter=250, learning_rate=50,
+               n_components=n_embedding, random_state=0,
+               metric='precomputed').fit_transform(precomputed_X)
+
+    now = TSNE(verbose=1, perplexity=40, n_iter=250, learning_rate=50,
+               n_components=n_embedding, random_state=0, metric='mahalanobis',
+               metric_params={'V': np.cov(X.T)}).fit_transform(X)
+    assert_array_equal(ref, now)
+
+
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+def test_tsne_metric_params(method):
+    """Make sure that tha mahalanobis distance works with metric_params
+       properly set and it doesn't otherwise"""
+    random_state = check_random_state(0)
+    n_features = 10
+    n_embedding = 3
+    n_samples = 100
+    X = random_state.randn(n_samples, n_features)
+
+    # 1. check case metric='minkowski', p=2
+    precomputed_X = squareform(pdist(X, metric='minkowski', p=2), checks=True)
+    ref = TSNE(verbose=1, perplexity=40, n_iter=250, learning_rate=50,
+               n_components=n_embedding, random_state=0,
+               metric='precomputed', method=method
+               ).fit_transform(precomputed_X)
+
+    now = TSNE(verbose=1, perplexity=40, n_iter=250, learning_rate=50,
```

Comment:
this would be easier to read if you constructed TSNE once, then used `set_params` to only set the differences in set up without repeating other params.